### ISSUE TITLE
Handle extension ranges like: "SO:0000234|SO:0000655|TranscriptID"

### ISSUE THREAD

**Issue by kimrutherford** [pombase/canto#1007]

**Comment by kimrutherford** [comment]
Should this just be a free text range for now?  (see #1019)

**Comment by ValWood** [comment]
So is it difficult to be SO term and Gene ID simultaneously?
If so I think we can  use free text for a while (no different from what we do already).
Do you envisage a solution?

**Comment by kimrutherford** [comment]
> So is it difficult to be SO term and Gene ID simultaneously?

It's not too bad but it will take a bit of time.  I wasn't thinking it could be delayed if we wanted to get the new code into production quicker.

I'll have a deeper think about how to implement it.

**Comment by ValWood** [comment]
yes that can be delayed...

**Comment by mah11** [comment]
This affects almost half the configured FYPO extension patterns, including some heavily used ones like "protein level" and "protein localization" phenotypes.

**Comment by ValWood** [comment]
I seems this will need to happen before we go live.
Odd that it is so frequent (to me, as I rarely use SO terms....)

**Comment by mah11** [comment]
I think the SO IDs in the summary are just examples. From the test Canto behavior, I inferred that it's actually any set of pipe-delimited IDs in the extension range that are affected, and that's why it's a high priority for FYPO extensions.

**Comment by ValWood** [comment]
So far SO terms have only been used for 166 annotations for 33 FYPO terms?

I think in cases where SO terms were allowed, we could default to "genes" until Kim has time to figure out how. If anyone has something they cannot capture they can put  it in the feedback.

**Comment by ValWood** [comment]
I just saw you previous comment.

I think for the pipe separated ones here means different types of IDs, Kim is making these free text in the short term and then restricting. 

My understanding is when 2 different *types* of ID are allowed (PomBase +SO) this is a bit harder to implement.

**Comment by mah11** [comment]
Most of the affected FYPO extension config options do have multiple ID types, e.g. TranscriptID|SO:0000673 or ProteinID|PR:000000001

**Comment by ValWood** [comment]
Kim- for those which are
GeneID|SO:0001411 or ProteinID|PR:000000001
default to "GeneID" for going live
(because we  use SO/PR IDs very infrequently).

We will then flag this as high priority ticket  to follow up immediately, this will work for  99.6 % of phenotype annotations as a stop gap.

(unless of course you figure out a quick/easy way to do it)

**Comment by kimrutherford** [comment]
> for those which are GeneID|SO:0001411 or ProteinID|PR:000000001
> default to "GeneID" for going live

As a hack, I've changed the config parser to use just the first value in the range field, so for "GeneID|SO:0001411" it will treat it as "GeneID".  That is very temporary.

**Comment by kimrutherford** [comment]
> As a hack, I've changed the config parser to use just the first value in the range field

I'm made another slightly improvement: now that #1023 is implemented all ontology terms are used.  So that in this example:

    SO:0000234|SO:0000655|TranscriptID

The two ontology terms will be used to restrict the term name search (but the "TranscriptID" is still ignored.

**Comment by kimrutherford** [comment]
We need to discuss how the interface should work in the case.  Should we show the gene name dropdown and the ontology autocomplete box at the same time?  Do you next separate text to display beside each one?

**Comment by ValWood** [comment]
I am testing.
For this
GO:0006351      is_a    occurs_at       SO:0001865|SO:0001795   occurs at       0,1     user
I expected to see no option currently, should I?

**Comment by kimrutherford** [comment]
> I expected to see no option currently, should I?

The term name search is just one text field.  The autocomplete in this case is restricted to show only terms that are descendants of SO:0001865 or SO:0001795 (or those two terms themselves).

**Comment by ValWood** [comment]
yep got it...I wasnt typing the right thing

**Comment by ValWood** [comment]
Re: > We need to discuss how the interface should work in the case. Should we show the gene name dropdown and the ontology autocomplete box at the same time? Do you next separate text to display beside each one?

I prefer to show both so the user can select the option that is most relevant...(maybe one above the other separated by 
OR

Anyone else? Would this be odd?

**Comment by mah11** [comment]
Based on some FYPO examples, such as transcription regulation or RNA binding, I think yes, we should show the gene name dropdown and the ontology autocomplete box at the same time -- separated by a nice obvious "OR".

Possible text prompts

- for TranscriptID|SO:nnn

    gene transcript [pulldown]
    OR
    transcript type [autocomplete box]

- for ProteinID|PR:000000001

    protein encoded by [pulldown]
    OR
    protein isoform, modified form, etc. [autocomplete box]

**Comment by ValWood** [comment]
+1

**Comment by kimrutherford** [comment]
Any thoughts on how it should look in the configuration file?  Currently there is one column for the text to display.

**Comment by ValWood** [comment]
M.

**Comment by mah11** [comment]
I would utterly love it if this could be handled by the Canto code instead of the configuration file, because that's how it works now. The text in my previous suggestions was for the step after the one that has text in the config files; i.e. it's for the box where you choose the ID part of relation(ID), whereas the config files at present only specify what gets displayed for the relation part.

If we do push it onto the config file, I guess it'll need a new column (argh argh argh) and header changes - e.g. existing "Canto display text" renamed "Canto relation display text" and new column "Canto ID prompt text".

We will also need a way to distinguish two cases:

- allow one or the other (not both) per extension
- allow one of each per extension

At present, I've made a sort of a de-facto distinction in the FYPO config file by putting each one-not-both case on one line, and using separate lines for allow-one-of-each (proper handling pending bug fix in #1008).

Furthermore, for the don't-allow-both case, we will need to specify which ID prompt goes with which relation. This bit scares me, because if we leave it on one line as it is now, we would have to make sure the ID type and accompanying text go on the line in theright order, and that's begging to be screwed up. On the other hand, if we put them on separate lines, then we'd lose one vs. many lines as a way to distinguish allow-one-of-each from allow-both.

**Comment by ValWood** [comment]
Hi Kim,

Let us know if you need anything else from us to proceed with this one (and poss follow up with M. Monday or Tuesday?)
Midori will test phenotype stuff when this one is done.

Val

**Comment by kimrutherford** [comment]
> Let us know if you need anything else from us to proceed with this one

I think I'll need to Skype call about this as I'm find it a bit confusing.

Could we chat on Skype this week Midori?  Monday, Wednesday or Thursday morning would be best for me.

**Comment by mah11** [comment]
Thoughts from the Kim/Midori skype chat today:

For proteins, make PR:nnn IDs admin-only, and add a note saying to pester curators for things we would use PR:nnn for (possible text: please add a comment if you want to indicate a specific modified form of the protein) 

Admin curators will then get two choices, but we'll know to use just one.

M to update FYPO config file accordingly.

For transcripts, use radio buttons for the "gene transcript" and "transcript type" options. Kim will put some prompt text in autocomplete boxes (for ontology terms generally, not just for transcripts) for us to try.

**Comment by mah11** [comment]
FYPO config changes done.

**Comment by mah11** [comment]
RE another small thing we chatted about: I've requested at LTR transcript term from SO, so we can just point to the SO 'transcript' term for FYPO:0000824. In the meantime I've made SO:0000286 admin-only.

https://github.com/The-Sequence-Ontology/SO-Ontologies/issues/341

**Comment by kimrutherford** [comment]
> Kim will put some prompt text in autocomplete boxes

I've done this.  I think it makes it clearer in some case but not in others.

**Comment by kimrutherford** [comment]
>  I think it makes it clearer in some case but not in others.

Here's an example of an extension from BP:

     rDNA_replication_fork_barrier or mating_type_region_replication_fork_barrier or tRNA_gene ...

It would probably help to turn the underscores to spaces but it's still too long for the autocomplete box.  We might need a new plan.

**Comment by mah11** [comment]
hmm. Maybe the first (or shortest) one or two, and then "etc. ..."?

**Comment by ValWood** [comment]
I thought these would be individual items in a drop down. Am I missing something?

**Comment by kimrutherford** [comment]
We can't always put everything in a dropdown if the range is large.  The example I gave above is easy because there aren't many child terms.  But for some things it's a bit busy:
SO:0000296|SO:0001272|SO:0001914|SO:0002021|SO:0002020|SO:0000141|SO:0000186|SO:0000101|SO:0000286|SO:0001263|SO:0001846|SO:0001871|SO:0001846|SO:0001496|SO:0001003|SO:0001843
(From GO_CC_A_E_config)

It's a long list and some of those have child terms.

The current code just shows one autocomplete box where you can type the name of any of the terms configured in the range, or any child term.

I'm hoping to fix things though so that if you have a small number of possible terms (maybe 10-15) it will automatically show a dropdown instead (#1022).

**Comment by ValWood** [comment]
Ah thats what I understood. Small number drop down, large number autocomplete....

**Comment by kimrutherford** [comment]
For some terms this isn't working so well.  eg. FYPO:0002333:

![screenshot 2015-12-14 11 14 06](https://cloud.githubusercontent.com/assets/90474/11769766/e393edea-a253-11e5-9b90-db983ffe3022.png)

**Comment by mah11** [comment]
I'm not sure why there are four "assayed protein ..." links showing up; 3 of them are obvious in the config file, and I'll have to try to track down the 4th.
In the meantime, I can adjust the display text for the admin-only entries for FYPO:0002333. That should take care of two of them.

**Comment by ValWood** [comment]
This should may be be split into multiple ticket:

The first part is to  to handle extensions with multiple times of extensions
i.e  both  SO terms and  and transcripts

Other issues
If you have a checkbox decide on of the autocomplete fields it is a bit unclear what to put in the box. It should put a placeholder text (in grey)
This bit was a bit unclearer so Kim will implement the first part and then we will reassess

Another issue
The part about "assayed protein", Kim will make a separate ticket for this

Kim is going to implement  the first part as described, and then we will decide what text is needed if any.

**Comment by kimrutherford** [comment]
> The part about "assayed protein", Kim will make a separate ticket for this

Done: #1047

**Comment by kimrutherford** [comment]
I've added radio buttons to the extension dialog box if is more than one range type for the relation.

Here's a screenshot.  It needs work on the text labels but it's something to talk about.  Could you have a quick try and think about how it should look?

![screenshot 2016-01-06 20 02 35](https://cloud.githubusercontent.com/assets/90474/12136901/034bd0f2-b4b1-11e5-983c-b3beb1b0e0d9.png)

**Comment by airvine** [comment]
Would it possible to have multiple values for the "canto display text" in the annotation extension config file, separated by | (vertical bar), such that each display value matched the appropriate range type?

In the example above, there is only one "canto display text" value, "assayed region", so this is displayed for every range type. I guess these values would need to be distinguished from the initial text displayed in the annotation extension page, so a separate column may be required.

**Comment by mah11** [comment]
HI Alistair,

Your question about the display text is a good one, and is basically the same thing (or at least part of it) as I was struggling with in my comment from Nov. 25th (https://github.com/pombase/canto/issues/1007#issuecomment-159610872 above). I got rid of some clashes for PomBase FYPO extensions by amending the configuration, but I don't think we've entirely worked out what will be the best way to deal with any remaining ones.

m

**Comment by kimrutherford** [comment]
>  don't think we've entirely worked out what will be the best way to deal with any remaining ones.

I agree.  It needs a bit of thought.

**Comment by ValWood** [comment]
I'm not sure where we are up to or what the problem is so I added this to Tuesday call to discuss

**Comment by airvine** [comment]
For PHI-base, I don't use FYPO, so I may not be fully understanding the problem, but the extension dialog box if is more than one range type for the relation.

FYPO:0003452	is_a	assayed_usingPR:000000001	assayed protein isoform	0,1	admin
FYPO:0004083	is_a	assayed_using	PR:000000001	assayed protein isoform	0,1	admin

would be better defined as...

FYPO:0003452|FYPO:0004083	is_a	assayed_using	PR:000000001	assayed protein isoform	0,1	admin
	is_a	assayed_using	PR:000000001	assayed protein isoform	0,1	admin

**Comment by mah11** [comment]
From a little while ago ...

> I've added radio buttons to the extension dialog box if is more than one range type for the relation.

I think the radio button basic appearance and behaviour is fine. Let's open a new ticket for any text changes, because details like that will get totally lost in this thread.

**Comment by kimrutherford** [comment]
> One problem is the display for different range types in the extension dialog box if there is more than one range type for the relation.

Yep, that needs discussion.  The config file is already a bit hairy so I'd like to keep it simple if we can.

> would be better defined as...
>      FYPO:0003452|FYPO:0004083   is_a    assayed_using   PR:000000001    assayed protein isoform 0,1 adm

Yep, that would help make the config file tidier.  I've made a separate issue for that: #1071

Sometimes though it would be nice if Canto worked it out and fixed it.  I mean: if a term gets identical configuration from two different ancestors, it should display that config just once.

**Comment by ValWood** [comment]
I think this is all good. I haven't re-read everything in the ticket, so if anything is not working, new ticket....

**Comment by kimrutherford** [comment]
> so if anything is not working, new ticket....

There are still problems.  I've made a new issue: #1079
